bakerytest/bakerytest.go

```diff
@@ -127,3 +130,131 @@ func (d *Discharger) PublicKeyForLocation(loc string) (*bakery.PublicKey, error)
 	}
 	return nil, bakery.ErrNotFound
 }
+
+type discharge struct {
+	cavId string
+	c     chan error
+}
+
+// LoginDischarger is a Discharger that always requires interraction to
+// complete the discharge.
+type LoginDischarger struct {
+	Discharger
+	Mux *http.ServeMux
+
+	// mu protects the following fields.
+	mu      sync.Mutex
+	waiting map[string]discharge
+	id      int
+}
+
+// NewLoginDischarger returns a new LoginDischarger which uses
+// loginHandler to handle the login interraction. The login is completed
+// by calling FinishLogin.
+//
+// If locator is non-nil, it will be used to find public keys
+// for any third party caveats returned by the checker.
+//
+// Calling this function has the side-effect of setting
+// InsecureSkipVerify in http.DefaultTransport.TLSClientConfig
+// until all the dischargers are closed.
+func NewLoginDischarger(locator bakery.PublicKeyLocator, loginHandler http.Handler) *LoginDischarger {
+	ld := &LoginDischarger{
+		Mux:     http.NewServeMux(),
+		waiting: map[string]discharge{},
+	}
+	ld.Mux.Handle("/login", loginHandler)
+	ld.Mux.Handle("/wait", http.HandlerFunc(ld.wait))
+	server := httptest.NewTLSServer(ld.Mux)
+	svc, err := bakery.NewService(bakery.NewServiceParams{
+		Location: server.URL,
+		Locator:  locator,
+	})
+	if err != nil {
+		panic(err)
+	}
+	httpbakery.AddDischargeHandler(ld.Mux, "/", svc, ld.checker)
+	startSkipVerify()
+	ld.Discharger = Discharger{
+		Service: svc,
+		server:  server,
+	}
+	return ld
+}
+
+func (d *LoginDischarger) checker(req *http.Request, cavId, cav string) ([]checkers.Caveat, error) {
+	d.mu.Lock()
+	id := fmt.Sprintf("%04x", d.id)
+	d.id++
+	d.waiting[id] = discharge{cavId, make(chan error, 1)}
+	d.mu.Unlock()
+	visitURL := fmt.Sprintf("%s/login?waitid=%s", d.Discharger.server.URL, id)
+	waitURL := fmt.Sprintf("%s/wait?waitid=%s", d.Discharger.server.URL, id)
+	return nil, httpbakery.NewInteractionRequiredError(visitURL, waitURL, nil, req)
+}
+
+func (d *LoginDischarger) wait(w http.ResponseWriter, r *http.Request) {
+	r.ParseForm()
+	d.mu.Lock()
+	discharge, ok := d.waiting[r.Form.Get("waitid")]
+	d.mu.Unlock()
+	if !ok {
+		httprequest.WriteJSON(w, http.StatusBadRequest, httpbakery.Error{
+			Message: fmt.Sprintf("invalid waitid %q", r.Form.Get("waitid")),
+		})
+		return
+	}
+	defer func() {
+		d.mu.Lock()
+		delete(d.waiting, r.Form.Get("waitid"))
+		d.mu.Unlock()
+	}()
+	err := <-discharge.c
+	if err != nil {
+		herr, ok := err.(*httpbakery.Error)
+		if !ok {
+			herr = &httpbakery.Error{
+				Code:    "forbidden",
+				Message: err.Error(),
+			}
+		}
+		httprequest.WriteJSON(w, http.StatusForbidden, herr)
+		return
+	}
+	m, err := d.Service.Discharge(
+		bakery.ThirdPartyCheckerFunc(
+			func(cavId, caveat string) ([]checkers.Caveat, error) {
+				return nil, nil
+			},
+		),
+		discharge.cavId,
+	)
+	if err != nil {
```

Comment:
I'd also use httpbakery.ErrorToResponse here too. I guess we *could* just use httprequest's HandleErrors method and just return an error from this function.